iceoryx_posh/include/iceoryx_posh/internal/popo/client_impl.inl

```diff
@@ -0,0 +1,81 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_POSH_POPO_CLIENT_IMPL_INL
+#define IOX_POSH_POPO_CLIENT_IMPL_INL
+
+#include "iceoryx_posh/internal/popo/client_impl.hpp"
+
+namespace iox
+{
+namespace popo
+{
+template <typename Req, typename Res, typename BaseClientT>
+ClientImpl<Req, Res, BaseClientT>::ClientImpl(const capro::ServiceDescription& service,
+                                              const ClientOptions& clientOptions) noexcept
+    : BaseClientT(service, clientOptions)
+{
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+cxx::expected<Request<Req>, AllocationError> ClientImpl<Req, Res, BaseClientT>::loanUninitialized() noexcept
+{
+    auto result = port().allocateRequest(sizeof(Req), alignof(Req));
+    if (result.has_error())
+    {
+        return cxx::error<AllocationError>(result.get_error());
+    }
+    auto requestHeader = result.value();
+    auto payload = mepoo::ChunkHeader::fromUserHeader(requestHeader)->userPayload();
+    auto request = cxx::unique_ptr<Req>(reinterpret_cast<Req*>(payload), m_requestDeleter);
+    return cxx::success<Request<Req>>(Request<Req>{std::move(request), *this});
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+template <typename... Args>
+cxx::expected<Request<Req>, AllocationError> ClientImpl<Req, Res, BaseClientT>::loan(Args&&... args) noexcept
+{
+    return std::move(
+        loanUninitialized().and_then([&](auto& request) { new (request.get()) Req(std::forward<Args>(args)...); }));
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+void ClientImpl<Req, Res, BaseClientT>::send(Request<Req>&& request) noexcept
+{
+    // take the ownership of the chunk from the Request to transfer it to `sendRequest`
+    auto payload = request.release();
+    auto* requestHeader = static_cast<RequestHeader*>(mepoo::ChunkHeader::fromUserPayload(payload)->userHeader());
+    port().sendRequest(requestHeader);
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+cxx::expected<Response<const Res>, ChunkReceiveResult> ClientImpl<Req, Res, BaseClientT>::take() noexcept
+{
+    auto result = port().getResponse();
+    if (result.has_error())
+    {
+        return cxx::error<ChunkReceiveResult>(result.get_error());
+    }
+    auto responseHeader = result.value();
+    auto payload = mepoo::ChunkHeader::fromUserHeader(responseHeader)->userPayload();
+    auto response = cxx::unique_ptr<const Res>(static_cast<const Res*>(payload), m_responseDeleter);
+    return cxx::success<Response<const Res>>(Response<const Res>{std::move(response)});
```

Comment:
Why is there a need to construct a `Response` from an `unique_ptr`? In theory it could be constructed from `payload` and some `responseDeleter` directly, could it not? Internally it would probably use a `unique_ptr` to not reinvent the wheel though (as it uses RAII with custom deleter).

Just thinking about the design here ...

Reply:
Good point. It was modeled as per the Sample, which also takes a `unique_ptr`. I also have to check which implications this has but most probably could change it like you suggested.

Reply:
If you don't mind, I would keep this as is since it is an internal API and I'm not quite sure how much code needs to be changed since this part is shared with the `SmartChunk` which is the base of the `Sample`